internal/util/serdeio.go

```diff
@@ -35,6 +35,11 @@ func ReadJSONOrYAML(src io.Reader, dest proto.Message) error {
 	return d.decode(dest)
 }
 
+func IsJSON(src []byte) bool {
```

Comment:
If I'm being pedantic,

```yaml
{ foo: [bar, baz] }
```

is valid YAML but not valid JSON.

I think the parser should be aware if it is parsing YAML or JSON based on the file extension of the input, and valid YAML in a `.json` file should be rejected (presumably someone using `.json` policies is doing so because they want them to be easily machine-parseable, so it would be better to be strict).

Reply:
Good point. The problem is that the parser works off of an `io.Reader` because I wanted it to be generic and work with any byte stream. The file extension information gets lost because of that. Maybe the signature should be changed to let the caller define whether the stream is YAML or JSON 🤔

Reply:
> Maybe the signature should be changed to let the caller define whether the stream is YAML or JSON 🤔

Sounds good to me!

Reply:
So, even if the caller provides a hint about the kind of bytes we are dealing with, the underlying parser doesn't provide any indication about whether the contents are YAML or JSON. Without that information, I don't think we can enforce that `.json` files contain only JSON and so forth.

The only reason for the `IsJSON` function here is because the detection of quoted strings in YAML gives false positives for single-line JSON content. Furthermore, in the current version of Cerbos, we do a [similar check and disregard the extension](https://github.com/cerbos/cerbos/blob/8db9898212a72a52c840a753addcec0635d5cbd0/internal/util/serdeio.go#L38-L48) so if we change that behaviour it would be a breaking change as well. Therefore, I am inclined to keep it as it is.

Reply:
Ok, fair enough!